### ISSUE TITLE
Peek operation for queues

### ISSUE THREAD

**Issue by danijar** [tensorflow/tensorflow#7880]
Peeking queues would be a useful addition for multi threaded TensorFlow applications. Right now there is no way to look at the first element without affecting the queue. We could empty a `tf.FIFOQueue`, look at the first element, and enqueue all elements in order. But that is not safe when other threads operate on the queue at the same time (see locking mechanisms #6360).

Example use case is an environment class in reinforcement learning, that buffers the next few time steps in a queue. The agent should be able to look at the current observation and reward at any time. Executing a step operation should remove the current time step so the next one becomes visible to the agent.

**Comment by yaroslavvb** [comment]
There was an internal discussion @suharshs / @mrry once about this (to mitigate the unexpected behavior of summary ops consuming training data). I'm hazy on details, maybe @mrry remembers, there were some design/implementation difficulties in supporting peek

**Comment by danijar** [comment]
Do you think it's possible to build a workaround from existing TensorFlow operations?

**Comment by yaroslavvb** [comment]
1. Work-around is possible by having all accesses to a queue go through buffer. Your would have a "queue" object which consists of tf.FIFOQueue and tf.Variable, and the top object of the "queue" is stored in the tf.Variable. You can then implement "peek" that returns value of the variable, then and "dequeue" which overwrites it with value from the FIFOQueue

2. Native support would be useful since `.peek` is a common abstraction

**Comment by danijar** [comment]
Thanks, I will use that. Similarly, having a `clear()` operation would be useful instead of checking for non emptiness and using `dequeue_up_to()` with the queue capacity.

**Comment by danijar** [comment]
@yaroslavvb Actually, I implemented your idea of holding the first element in a variable. Unfortunately, it doesn't seem to work. The problem might be that, unlike the queue, the variable is not safe against concurrent access. Here is [an example gist](https://gist.github.com/danijar/fe98e758700865124b109bd1ee85f5aa).

**Comment by yaroslavvb** [comment]
@danijar your gist is non-deterministic because different valid execution orders produce different reuslts. Consider following block

```
    dummy = tf.cond(tf.equal(self._size, 0), first, other)
    with tf.control_dependencies([self._size.assign_add(1)]):
      return tf.identity(dummy)
```

Since `dummy=tf.cond` is created outside of the `control_dependencies` block, `cond` can run either before or after `assign_add`. You can make it deterministic by moving it inside the block

```
    with tf.control_dependencies([self._size.assign_add(1)]):
      dummy = tf.cond(tf.equal(self._size, 0), first, other)
      return tf.identity(dummy)
```

**Comment by yaroslavvb** [comment]
possibly relevant -- https://github.com/tensorflow/tensorflow/issues/7951

**Comment by danijar** [comment]
Thank you. However, then the condition changes. I'd either need to compare the size to 1 instead which makes the code confusing, or add a second `tf.control_dependencies()` around everything to ensure that `dummy` was gets executed before. In any way, it would be quite helpful to have an existing `.peek()` on queues.

**Comment by georgekrasenov** [comment]
+1 for .peek()

**Comment by suharshs** [comment]
Hi all,
A peek operation on concurrent queues is inherently racy since there can be multiple consumers (one consumer peeks while the other consumer takes). It seems ill-advised to add a method to queues that only works in the single consumer case.
Additionally, peek may be defined differently (or add a lot of complexity) for non FIFO queues.
If you share your use cases, we may be able to try to find more workarounds.
Thanks!

**Comment by danijar** [comment]
@suharshs The use case is reinforcement learning with the environment ticking a few time steps ahead of the agent to allow to parallelize the two. Both environment and agent fully live in graph and communicate via queues. The agent sends actions with a queue to the environment, which gets consumed by a queue runner that steps the environment. Moreover, the environment provides queues containing observations, rewards, and done flags that the agent should `peek()` at.

**Comment by danijar** [comment]
@suharshs Do you have a solution for in-graph reinforcement learning?

**Comment by suharshs** [comment]
I do not, but this is definitely something that people are interested in and we plan to look into. I will close this issue though for now, and keep you posted on in-graph RL solutions
